pallets/creditcoin/src/tests/cleanup.rs

```diff
@@ -0,0 +1,289 @@
+				crate::CleanupState::<TestRuntime>::put(state);
+			}
+		});
+		t.commit_all().unwrap();
+		t
+	}
+}
+
+#[extend::ext]
+impl TestExternalities {
+	fn run<F: FnOnce() -> R, R>(&mut self, f: F) -> R {
+		let res = self.execute_with(f);
+		self.commit_all().unwrap();
+		res
+	}
+	fn then_run(&mut self, f: impl FnOnce()) -> &mut TestExternalities {
+		self.run(f);
+		self
+	}
+}
+
+type AskOrders = crate::AskOrders<TestRuntime>;
+type BidOrders = crate::BidOrders<TestRuntime>;
+type Offers = crate::Offers<TestRuntime>;
+
+// has to be called with externalities in scope
+fn cleanup(block: u64) -> Weight {
+	Creditcoin::on_initialize(block)
+}
+
+fn cleanup_state() -> Option<StorageCleanupState<BlockNumber>> {
+	crate::CleanupState::<TestRuntime>::get()
+}
+
+const LIMIT: usize = mock::CLEANUP_LIMIT as usize;
+
+#[test]
+fn cleans_up_asks() {
+	let count = LIMIT / 2;
+	let exp = 5;
+
+	StorageBuilder::new().ask_orders(count, exp).finish().execute_with(|| {
+		assert_eq!(crate::AskOrders::<TestRuntime>::iter().count(), count);
+
+		cleanup(exp);
+
+		assert_eq!(crate::AskOrders::<TestRuntime>::iter().count(), 0);
+		assert_eq!(cleanup_state(), Some(StorageCleanupState::new(exp + 1)));
+	});
+}
+
+fn weak_bounded<S: Get<u32>>(v: Vec<u8>) -> WeakBoundedVec<u8, S> {
+	WeakBoundedVec::force_from(v, None)
+}
+
+macro_rules! impl_get_partial_key {
+	($($storage: ty),*) => {
+		$(
+			impl GetPartialKey for $storage {
+				fn partial_key(first: u64) -> Vec<u8> {
+					<$storage>::storage_double_map_final_key1(first)
+				}
+			}
+		)*
+	};
+}
+
+trait GetPartialKey {
+	fn partial_key(exp: u64) -> Vec<u8>;
+}
+
+impl_get_partial_key!(AskOrders, BidOrders, Offers);
+
+fn cursor_for<Storage: GetPartialKey>(
+	first: u64,
+) -> Option<WeakBoundedVec<u8, ConstU32<{ StorageItemCleanupState::<BlockNumber>::MAX_CURSOR_LEN }>>>
+{
+	Some(weak_bounded(Storage::partial_key(first)))
+}
+
+#[test]
+fn cleans_up_asks_to_limit() {
+	let count = LIMIT * 2;
+	let exp = 5;
+
+	StorageBuilder::new()
+		.ask_orders(count, exp)
+		.finish()
+		.then_run(|| {
+			assert_eq!(AskOrders::iter().count(), count);
+
+			cleanup(exp);
+
+			assert_eq!(AskOrders::iter().count(), count - LIMIT);
+			assert_eq!(
+				cleanup_state(),
+				Some(StorageCleanupState {
+					ask_orders: StorageItemCleanupState {
+						on_block: exp,
+						cursor: cursor_for::<AskOrders>(exp)
+					},
+					..StorageCleanupState::new(exp)
+				})
+			);
+		})
+		.then_run(|| {
+			assert_eq!(AskOrders::iter().count(), count - LIMIT);
+
+			cleanup(exp + 1);
+
+			assert_eq!(AskOrders::iter().count(), 0);
+			assert_eq!(
+				cleanup_state(),
+				Some(StorageCleanupState {
+					ask_orders: StorageItemCleanupState::new(exp + 1),
+					..StorageCleanupState::new(exp)
+				})
+			);
+		})
+		.then_run(|| {
+			assert_eq!(AskOrders::iter().count(), 0);
+
+			cleanup(exp + 2);
+
+			assert_eq!(AskOrders::iter().count(), 0);
+			assert_eq!(cleanup_state(), Some(StorageCleanupState::new(exp + 3)));
```

Comment:
The end state that we're asserting on is the point where you'll resume work next block. The current block is `exp + 2`, and since there were no ask orders we're done with the work for `exp + 2`. So next block you would be cleaning up entries that are at current block plus one = `exp + 3`.